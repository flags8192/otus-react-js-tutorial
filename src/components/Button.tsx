import React from 'react';
import styled from '@emotion/styled';
import { FC } from 'react';

const Root = styled.button`
  height: 30px;
  margin: auto 20px;
`;

export interface IButton {
  title?: string;
  onClick?: () => void;
}
const Button: FC<IButton> = ({ title = 'Button', onClick }) => {
  return (
    <Root role="button" onClick={onClick}>
      {title}
    </Root>
  );
};

export default Button;
