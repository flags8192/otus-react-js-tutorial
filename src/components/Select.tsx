import React from 'react';
import styled from '@emotion/styled';
import { FC } from 'react';

const Root = styled.select`
  height: 30px;
  margin: auto 20px;
`;

export interface ISelect {
  onSelect?: (level: React.ChangeEvent<HTMLSelectElement>) => void;
}
const Select: FC<ISelect> = ({ onSelect }) => {
  return (
    <Root role="select" name="select" onChange={onSelect}>
      <option role="option" value="1" defaultValue="1">
        Падаван
      </option>
      <option role="option" value="2">
        Джедай
      </option>
      <option role="option" value="3">
        Мастер-джедай
      </option>
      <option role="option" value="4">
        Член Совета Силы
      </option>
    </Root>
  );
};

export default Select;
