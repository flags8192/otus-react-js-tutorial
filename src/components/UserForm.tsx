import React, { PureComponent } from 'react';
import { Formik, Field, Form } from 'formik';
import styled from '@emotion/styled';

const Root = styled.div`
  label {
    margin-right: 5px;
  }

  button {
    margin: 5px;
  }

  input {
    margin-right: 30px;
  }

  form {
    font-size: 14px;
    margin-bottom: 10px;
  }
`;

const UserInfo = styled.label`
  opacity: 0.5;
  margin-left: 25px;
`;

export interface IUserFormProps {
  firstName: string;
  lastName: string;
  onSubmit: (values: IUserData) => void;
}

export type IUserData = Omit<IUserFormProps, 'onSubmit'>;

export class UserForm extends PureComponent<IUserFormProps, IUserData> {
  constructor(props: IUserFormProps) {
    super(props);

    this.state = {
      firstName: props.firstName,
      lastName: props.lastName,
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.getFio = this.getFio.bind(this);
  }

  handleSubmit() {
    this.props.onSubmit({ ...this.state });
  }

  getFio() {
    return `${this.state.firstName} ${this.state.lastName}`;
  }

  handleChange(e: React.ChangeEvent<HTMLInputElement>) {
    const { id, value } = e.currentTarget;
    this.setState((state) => ({ ...state, [id]: value }));
  }

  render() {
    return (
      <Root role="userForm">
        <Formik initialValues={{ ...this.state }} onSubmit={this.handleSubmit}>
          <Form>
            <label htmlFor="firstName">Имя</label>
            <Field
              id="firstName"
              name="firstName"
              placeholder="Введите ммя"
              value={this.state.firstName}
              onChange={this.handleChange}
            />

            <label htmlFor="lastName">Фамилия</label>
            <Field
              id="lastName"
              name="lastName"
              placeholder="Введите фамилию"
              value={this.state.lastName}
              onChange={this.handleChange}
            />

            <button type="submit">Сохранить</button>

            <UserInfo>{this.getFio()}</UserInfo>
          </Form>
        </Formik>
      </Root>
    );
  }
}
