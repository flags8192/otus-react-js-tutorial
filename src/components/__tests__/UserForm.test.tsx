import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { UserForm } from '@src/components/UserForm';
import userEvent from '@testing-library/user-event';

const fio = {
  firstName: 'Иван',
  lastName: 'Иванов',
};

describe('UserForm', () => {
  it('Render <UserForm>', () => {
    const handleSubmit = jest.fn();
    const { asFragment } = render(<UserForm onSubmit={handleSubmit} {...fio} />);
    expect(asFragment()).toMatchSnapshot();
    const userForm = screen.getByRole(/userForm/gi);
    expect(userForm).toBeInTheDocument();
    const firstName = screen.getByDisplayValue(/Иван/gi);
    expect(firstName).toBeInTheDocument();
    const lastName = screen.getByDisplayValue(/Иванов/gi);
    expect(lastName).toBeInTheDocument();
  });

  it('Change input <UserForm>', async () => {
    const handleSubmit = jest.fn();
    render(<UserForm onSubmit={handleSubmit} {...fio} />);
    const firstName = screen.getByLabelText(/Имя/i);
    userEvent.clear(firstName);
    userEvent.type(firstName, 'Петр');
    const lastName = screen.getByLabelText(/Фамилия/i);
    userEvent.clear(lastName);
    userEvent.type(lastName, 'Петров');

    userEvent.click(screen.getByRole('button', { name: /Сохранить/i }));

    await waitFor(() =>
      expect(handleSubmit).toHaveBeenCalledWith({
        firstName: 'Петр',
        lastName: 'Петров',
      })
    );
  });
});
