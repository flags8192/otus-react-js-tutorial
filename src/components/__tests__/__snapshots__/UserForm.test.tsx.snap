// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm Render <UserForm> 1`] = `
<DocumentFragment>
  <div
    class="css-nga5lq"
    role="userForm"
  >
    <form
      action="#"
    >
      <label
        for="firstName"
      >
        Имя
      </label>
      <input
        id="firstName"
        name="firstName"
        placeholder="Введите ммя"
        value="Иван"
      />
      <label
        for="lastName"
      >
        Фамилия
      </label>
      <input
        id="lastName"
        name="lastName"
        placeholder="Введите фамилию"
        value="Иванов"
      />
      <button
        type="submit"
      >
        Сохранить
      </button>
      <label
        class="css-12vxlcv"
      >
        Иван Иванов
      </label>
    </form>
  </div>
</DocumentFragment>
`;
